vms/platformvm/validators/manager.go

```diff
@@ -0,0 +1,303 @@
+// Copyright (C) 2019-2022, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package validators
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ava-labs/avalanchego/cache"
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/validators"
+	"github.com/ava-labs/avalanchego/utils/constants"
+	"github.com/ava-labs/avalanchego/utils/math"
+	"github.com/ava-labs/avalanchego/utils/timer/mockable"
+	"github.com/ava-labs/avalanchego/utils/window"
+	"github.com/ava-labs/avalanchego/vms/platformvm/config"
+	"github.com/ava-labs/avalanchego/vms/platformvm/metrics"
+	"github.com/ava-labs/avalanchego/vms/platformvm/state"
+	"github.com/ava-labs/avalanchego/vms/platformvm/txs"
+)
+
+const (
+	validatorSetsCacheSize        = 64
+	maxRecentlyAcceptedWindowSize = 256
+	recentlyAcceptedWindowTTL     = 5 * time.Minute
+)
+
+var (
+	_ validators.State = (*manager)(nil)
+
+	ErrMissingValidatorSet = errors.New("missing validator set")
+)
+
+// P-chain must be able to provide information about validators active
+// at different heights. [QueryManager] interface encapsulates all the machinery
+// to achieve this.
+type QueryManager interface {
+	validators.State
+
+	GetValidatorIDs(subnetID ids.ID) ([]ids.NodeID, bool)
+}
+
+// Manager interface adds to QueribleSet the ability to blocks IDs
+// to serve GetMinimumHeight
+type Manager interface {
```

Comment:
I think we don't need this interface, we can have an interface with only Track that is implemented by the same struct.
We could also rename Track to OnAccept and name the interface AcceptedBlockTracker or similar.

Reply:
I think Track sucks and we should find a better name. OnAccept is a reasonble suggestion and I have implemented it.
On the other hand, I don't really like the idea of  reducing Manager interface to `AcceptedBlockTracker`.
This manager is a pretty big class doing a bunch of not-so-related things, which are together only because there are listed under the same `validators.State`. It seems to me  that `AcceptedBlockTracker` goes in the direction of slicing off one of this things which may be the right thing to do here. Only if we go there, we should probably first slice `validators.State` and then extract the OnAccept interface.